### PR TITLE
Add IMPALA algorithm

### PR DESCRIPTION
**Pull request by zenghsh3**
- add impala algorithm
- add some common tools

### CODE REVIEW

File: parl/utils/window_stat.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = ['WindowStat']
+
+import numpy as np
+
+
+class WindowStat(object):
```
Comment (TomorrowIsAnOtherDay) [inline parl/utils/window_stat.py:20]: comments are needed here.

File: parl/utils/utils.py [hunk 2]
```diff
@@ -13,9 +13,11 @@
 # limitations under the License.
 
 import sys
+import numpy as np
 
 __all__ = [
-    'has_func', 'action_mapping', 'to_str', 'to_byte', 'is_PY2', 'is_PY3'
+    'has_func', 'action_mapping', 'to_str', 'to_byte', 'is_PY2', 'is_PY3',
+    'MAX_INT32', 'np_softmax', 'np_cross_entropy'
```
Comment (TomorrowIsAnOtherDay) [inline parl/utils/utils.py]: This files provide some basic python functions that are some functional functions, not computational functions. `action_mapping`, `np_softmax`, `np_crsoo_entropy` are supposed to be moved elsewhere.

File: examples/IMPALA/impala_config.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    'env_num': 5,
+    'sample_batch_steps': 50,
+
+    #==========  learner config ==========
+    'train_batch_size': 1000,
+    'learner_queue_max_size': 16,
+    'sample_queue_max_size': 8,
+    'gamma': 0.99,
+
+    # steps of optimizer
```
Comment (TomorrowIsAnOtherDay) [inline examples/IMPALA/impala_config.py]: learning rate adjustment schedule: (train_step, learning_rate)

File: examples/IMPALA/impala_config.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    'lr_scheduler': [(0, 0.001), (20000, 0.0005), (40000, 0.0001)],
+
+    # steps of optimizer
```
Comment (TomorrowIsAnOtherDay) [inline examples/IMPALA/impala_config.py]: coefficient of policy entropy adjustment schedule: (train_step,  coefficient)

File: parl/algorithms/impala/impala.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+                 E.g. [B, C, H, W] in atari.
+        """
+        logits = self.model.policy(obs)
+        probs = layers.softmax(logits)
+
+        predict_actions = layers.argmax(probs, 1)
+
+        return predict_actions
+
+    def get_params(self):
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/impala/impala.py]: Fetching parameters from the agent is a frequently used function since we have to export the model's parameters from time to time, how about moving this function into our framework?

File: parl/algorithms/impala/impala.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        return self.model.get_params()
+
+    def set_params(self, params, gpu_id):
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/impala/impala.py]: Assigning model parameters for an agent is a frequently used function since we have to export the model's parameters from time to time, how about moving this function into our framework?

File: parl/env/atari_wrappers.py [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+    def _get_ob(self):
+        assert len(self.frames) == self.k
+        return np.concatenate(self.frames, axis=2)
+
+
+def wrap_deepmind(env, dim=84, framestack=True):
+    """Configure environment for DeepMind-style Atari.
+
+    Note that we assume reward clipping is done outside the wrapper.
```
Comment (TomorrowIsAnOtherDay) [inline parl/env/atari_wrappers.py]: Reward clipping had been done at line 272 by using `ClipRewardEnv`

File: examples/IMPALA/train.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+from learner import Learner
+
+
+def main(config):
+    learner = Learner(config)
+
+    try:
+        while True:
+            start = time.time()
+            learner.step()
```
Comment (TomorrowIsAnOtherDay) [inline examples/IMPALA/train.py]: Calling step here is trivial since you have called it at line 27.

File: examples/IMPALA/actor.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            env = wrap_deepmind(env, dim=config['env_dim'])
+            self.envs.append(env)
+        self.vector_env = VectorEnv(self.envs)
+
+        self.obs_batch = self.vector_env.reset()
+
+        obs_shape = env.observation_space.shape
+        # conv layer of fluid only support 'NCHW' input format now,
+        # we will convert 'NHWC' format to 'NCHW' format.
+        if config['obs_format'] == 'NHWC':
```
Comment (TomorrowIsAnOtherDay) [inline examples/IMPALA/actor.py]: As we are running the server in python where there is a global lock, we should remove as much as computation into clients. Maybe we could do things like transormation at the client side next time.

File: parl/utils/window_stat.py (hunk 1 above)
Comment (zenghsh3) [reply parl/utils/window_stat.py:20]: done

File: parl/utils/utils.py (hunk 2 above)
Comment (zenghsh3) [reply parl/utils/utils.py]: np_softmax and np_cross_entropy have been put in np_utils; action_mapping will be fixed in future pr.

File: examples/IMPALA/impala_config.py (hunk 3 above)
Comment (zenghsh3) [reply examples/IMPALA/impala_config.py]: done

File: parl/algorithms/impala/impala.py (hunk 5 above)
Comment (zenghsh3) [reply parl/algorithms/impala/impala.py]: done

File: examples/IMPALA/impala_config.py (hunk 4 above)
Comment (zenghsh3) [reply examples/IMPALA/impala_config.py]: done

File: parl/algorithms/impala/impala.py (hunk 6 above)
Comment (zenghsh3) [reply parl/algorithms/impala/impala.py]: done

File: parl/env/atari_wrappers.py (hunk 7 above)
Comment (zenghsh3) [reply parl/env/atari_wrappers.py]: done

File: examples/IMPALA/train.py (hunk 8 above)
Comment (zenghsh3) [reply examples/IMPALA/train.py]: done

File: examples/IMPALA/actor.py (hunk 9 above)
Comment (zenghsh3) [reply examples/IMPALA/actor.py]: done; add obs_format config in atari_wrappers.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
Excellent!
